### DISCUSSION TITLE
How to add/invite users from within the app

### DISCUSSION THREAD

**Discussion by dilly-sixmo** [supabase/supabase#9197] [Questions]
I would like for some of my logged in users (managers) to be able to create new user accounts (for employees) directly within the app.  I don't want them having to use the Supabase control panel.  The problem I have is that `inviteUserByEmail() ` is only callable when using the service_key, something I wish not to expose in my app.

How can I have a logged in user create/invite a new user without exposing the service_key?  Perhaps when they add a new row to the "Employees" table it could trigger the server to invite the user using the Employee's email field.  Is this possible?  Can the server call inviteUserByEmail() for me and how would I go about implementing that?

**Comment by silentworks** [comment]
You can use the `inviteUserByEmail` in a server side function and not expose it to the client at all. This is how I'm currently using it in my SvelteKit application which allows you to invite users to your app.

**Reply by dilly-sixmo** [reply]
When you say "server side function" are you referring to the Edge Functions within Supabase itself or are you talking about creating a separate API server (using node+express or whatever) to handle it (like a middleman)?

**Reply by saltcod** [reply]
Referring to a function that gets run server-side instead of client-side. Not sure what your stack is, but in NextJS these are called [api routes](https://nextjs.org/docs/api-routes/introduction)

You would basically call an api endpoint from the frontend, and that would run the inviteUserByEmail() function in the backend without exposing your service role key.

Running this function on the server instead of the client would be the gist no matter what stack you're using.

**Reply by dilly-sixmo** [reply]
Currently, my only server is supabase.  I have my client app made in Vue and it connects to supabase and that's it.  So where does this server-side function go?  Would I need to create a server just for this single purpose?  That seems a little ridiculous.  Can't supabase do this for me?  I haven't explored Supabase edge functions yet but is it possible to do it with that so I don't need this single-function middleman server in my app architecture?

**Reply by bubbleheadinc** [reply]
Where are you hosting your Vue app? I host mine in Netlify and use server functions there to do the invite user calls with the server key via an environment variable.

**Reply by dilly-sixmo** [reply]
I'm also on Netlify.  So I can set up the function on there?  If so, that would be a viable option.  I will check that out.  Thanks!

**Reply by bubbleheadinc** [reply]
Yes. You create a functions directory in the root of your app and when you push up your code, Netlify will deploy any functions in that directory. So your function may look something like this:

```
"use strict";
const { createClient } = require("@supabase/supabase-js");

const supabase = createClient(
  process.env.SUPABASE_URL,
  process.env.SUPABASE_SERVICE_ROLE_KEY
);

exports.handler = async (event, context) => {
  const payload = JSON.parse(event.body);
  const email = payload.email;

  try {
   
    const { data: user, error } = await supabase.auth.api.inviteUserByEmail(
        email,
        {
          redirectTo: 'https://your-website/any-path'      
        }
      );

      if (error) {
        throw error;
      }
      return {
        statusCode: 200,
        body: JSON.stringify({
          ...user,
        }),
      };
    
  } catch (error) {
    console.error("Something has gone wrong inviting user:", error);

    if (error.status) {
      return {
        statusCode: error.status,
        body: JSON.stringify({ error: error }),
      };
    } else {
      return {
        statusCode: 500,
        body: JSON.stringify({ error: error }),
      };
    }
  }
```

**Reply by dilly-sixmo** [reply]
Excellent!  I just set it up myself and tested it out and it's working perfectly.  Thanks so much.  This had me confused for some time but you made sense of it all for me.  Much appreciated.

**Comment by charbs-io** [comment]
@bubbleheadinc quick question, this is callable via an endpoint that netlify provides, but that means anyone can call that endpoint publicly. How do you go about securing this to ensure for example only logged in users with a specific role can call it? Perhaps check the headers for a JWT and look at the role assigned to that user?

**Reply by bubbleheadinc** [reply]
Right. Yeah you could do something with JWT headers, but there are some pitfalls there. Here is a Netlify Support response with a good explanation that may help.

https://answers.netlify.com/t/support-guide-how-to-apply-access-control-for-netlify-functions/46519

**Reply by charbs-io** [reply]
Appreciate your response, that's helpful. Thank you!

**Reply by bubbleheadinc** [reply]
@cayoub88 Circling back on this a bit...

Since you have to use a service role to call these Supabase functions, there needs to be an alternate way to verify the user calling the Netlify function as you eluded to. 

Supabase has Edge Functions now where you would get the requesting user by sending up the access token and then you would have the RLS policies kicking in to allow/deny. Although these Edge Functions are in Deno, which I'm not very familiar with. Here is the guide on it from Supabase: https://supabase.com/docs/guides/functions

Alternatively, if the functions remain in Netlify, you could send the access token in the header, then use https://github.com/auth0/node-jsonwebtoken or similar to verify the token is valid by checking the signature with the Supabase JWT secret key (found in Supabase settings). Once you know the token is valid, you could use the decoded result and make a call to Supabase ( `const { data, error } = await supabase.auth.admin.getUserById(userId)` ) to verify they are a user and then on successful response, check roles before allowing functions like inviteUser. I just did some tests with this and it seems to work well.

**Comment by devidevio** [comment]
If you don't want to spin up additional middleware you could go with the solution I mentioned here:
* https://github.com/supabase/supabase/discussions/6055#discussioncomment-4742764
